### PR TITLE
Cache unchanged files on dry run v2

### PR DESCRIPTION
**Pull request by dorrogeray**
Rework of https://github.com/rectorphp/rector-src/pull/3834

PR for issue https://github.com/rectorphp/rector/issues/7932

The idea is to cache unchanged files on --dry-run, so subsequent --dry-runs can be faster, while always producing same results.

### CODE REVIEW

File: packages/Parallel/WorkerRunner.php [hunk 1]
```diff
@@ -95,7 +95,7 @@ public function run(Encoder $encoder, Decoder $decoder, Configuration $configura
 
                     if ($errorAndFileDiffs[Bridge::SYSTEM_ERRORS] !== []) {
                         $this->invalidateFile($file);
-                    } elseif (! $configuration->isDryRun()) {
+                    } else if (! $configuration->isDryRun() || $errorAndFileDiffs[Bridge::FILE_DIFFS] === []) {
```
Comment (samsonasik) [inline packages/Parallel/WorkerRunner.php]: WorkerRunner only apply on parallel process, the non-parallel can happen in ApplicationFileProcessor
https://github.com/rectorphp/rector-src/blob/0d6800eea6b74352bd2dd16d7a2254c9398e4e98/src/Application/ApplicationFileProcessor.php#L136

Comment (dorrogeray) [reply packages/Parallel/WorkerRunner.php]: Thanks, I will update the PR to reflect this :+1:

Comment (samsonasik) [reply packages/Parallel/WorkerRunner.php]: `else if` can be `elseif` per coding standard we use.

**Review by samsonasik — APPROVED** [review]

**Comment by dorrogeray** [comment]
@samsonasik I have created a matrix to run the consecutive test with cache with dry run both with and without parallelization:

https://github.com/rectorphp/rector-src/actions/runs/5314204113/jobs/9621140293?pr=4281

**Comment by samsonasik** [comment]
@TomasVotruba e2e workflow update may need to be enabled:

```
 End to end test - e2e/consecutive-changes-with-cache Expected — Waiting for status to be repo
```

**Comment by TomasVotruba** [comment]
I'll review this now...

**Comment by TomasVotruba** [comment]
Thank you @dorrogeray :+1:

**Pull request merged**